File: app/src/main/java/com/capstone/project_niyakneyak/ui/main/adapter/CheckAlarmAdapter.java
package com.capstone.project_niyakneyak.ui.main.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.capstone.project_niyakneyak.R;
import com.capstone.project_niyakneyak.data.alarm_model.Alarm;
import com.capstone.project_niyakneyak.ui.main.listener.OnCheckedAlarmListener;

import java.util.Calendar;
import java.util.List;
import java.util.Locale;

/**
 *
 */
public class CheckAlarmAdapter extends RecyclerView.Adapter<CheckAlarmAdapter.ViewHolder> {
    private final OnCheckedAlarmListener onCheckedAlarmListener;
    private List<Alarm> alarms;

    public CheckAlarmAdapter(OnCheckedAlarmListener onCheckedAlarmListener){
        this.onCheckedAlarmListener = onCheckedAlarmListener;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder{
        private final TextView checkTitle;
        private final TextView checkRecursion;
        private final CheckBox checkBox;

        public ViewHolder(@NonNull View view) {
            super(view);
            checkTitle = view.findViewById(R.id.check_clock_time);
            checkRecursion = view.findViewById(R.id.check_clock_recursion);
            checkBox = view.findViewById(R.id.check_clock_checkbox);
        }

        public void bind(Alarm alarm, OnCheckedAlarmListener onCheckedAlarmListener){
            checkTitle.setText(String.format("%02d:%02d",alarm.getHour(),alarm.getMin()));
            if (alarm.isRecurring()) {
                checkRecursion.setText(alarm.getRecurringDaysText());
            } else {
                Calendar calendar = Calendar.getInstance();
                calendar.setTimeInMillis(System.currentTimeMillis());
                if(alarm.getHour() < calendar.get(Calendar.HOUR_OF_DAY)){
                    calendar.add(Calendar.DAY_OF_MONTH, 1);
                    checkRecursion.setText(String.format(Locale.KOREAN,"Next Day %02d/%02d",calendar.get(Calendar.MONTH) + 1, calendar.get(Calendar.DAY_OF_MONTH)));
                }
                else if(alarm.getHour() == calendar.get(Calendar.HOUR_OF_DAY)){
                    if(alarm.getMin() <= calendar.get(Calendar.MINUTE)){
                        calendar.add(Calendar.DAY_OF_MONTH, 1);
                        checkRecursion.setText(String.format(Locale.KOREAN, "Next Day %02d/%02d", calendar.get(Calendar.MONTH) + 1, calendar.get(Calendar.DAY_OF_MONTH)));
                    }
                    else checkRecursion.setText(String.format(Locale.KOREAN,"Today %02d/%02d", calendar.get(Calendar.MONTH) + 1, calendar.get(Calendar.DAY_OF_MONTH)));
                }
                else{
                    checkRecursion.setText(String.format(Locale.KOREAN,"Today %02d/%02d", calendar.get(Calendar.MONTH) + 1, calendar.get(Calendar.DAY_OF_MONTH)));
                }
            }
            checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> onCheckedAlarmListener.OnItemClicked(alarm, isChecked));
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_recycler_check_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Alarm alarm = alarms.get(position);
        holder.bind(alarm, onCheckedAlarmListener);
    }

    @Override
    public int getItemCount() {
        return alarms.size();
    }

    public void setAlarms(List<Alarm> alarms){
        this.alarms = alarms;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/capstone/project_niyakneyak/ui/main/adapter/CheckAlarmAdapter.kt
package com.capstone.project_niyakneyak.ui.main.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.capstone.project_niyakneyak.R
import com.capstone.project_niyakneyak.data.alarm_model.Alarm
import com.capstone.project_niyakneyak.ui.main.listener.OnCheckedAlarmListener

/**
 *
 */
class CheckAlarmAdapter(private val onCheckedAlarmListener: OnCheckedAlarmListener) :
    RecyclerView.Adapter<CheckAlarmAdapter.ViewHolder>() {
    private var alarms: List<Alarm>
    private var medsID: Long

    init {
        alarms = ArrayList()
        medsID = -1
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val checkTitle: TextView
        private val checkRecursion: TextView
        private val checkBox: CheckBox

        init {
            checkTitle = view.findViewById(R.id.check_clock_time)
            checkRecursion = view.findViewById(R.id.check_clock_recursion)
            checkBox = view.findViewById(R.id.check_clock_checkbox)
        }

        fun bind(medsID: Long, alarm: Alarm, onCheckedAlarmListener: OnCheckedAlarmListener) {
            checkTitle.text = String.format("%02d:%02d", alarm.hour, alarm.min)
            checkRecursion.text = alarm.daysText
            checkBox.setOnCheckedChangeListener { buttonView: CompoundButton?, isChecked: Boolean ->
                onCheckedAlarmListener.onItemClicked(medsID, alarm, isChecked)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_recycler_check_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val alarm = alarms!![position]
        holder.bind(medsID, alarm, onCheckedAlarmListener)
    }

    override fun getItemCount(): Int {
        return alarms!!.size
    }

    fun setAlarms(medsID: Long, alarms: List<Alarm>) {
        this.alarms = alarms
        this.medsID = medsID
        notifyDataSetChanged()
    }
}